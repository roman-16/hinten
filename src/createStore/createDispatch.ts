import get from 'just-safe-get';
import set from 'just-safe-set';
import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';
import { Reducer } from '../Store';

interface ReducerInfo {
  scope: string;
  reducer: Reducer;
}

export default (initialState: Record<string, unknown>, actions: Record<string, unknown>): [any, mitt.Emitter] => {
  const groupedReducer: Record<string, Array<ReducerInfo>> = {};
  const emitter = mitt();
  let currentState = initialState;

  return [
    mapValuesDeep(actions, ({ key, value, scope: actionScope }) => {
      if (!groupedReducer[key]) {
        groupedReducer[key] = [];
      }

      groupedReducer[key].push({
        scope: actionScope,
        reducer: (value as () => Reducer)(),
      });

      return (payload?: unknown) => {
        groupedReducer[key].forEach(({ scope, reducer }: ReducerInfo) => {
          // Ensure that the scope starts at least with the action-scope
          if (!scope.startsWith(actionScope)) return;

          const state = scope ? get(currentState, scope) : currentState;
          const newState = reducer(state, payload) as Record<string, unknown>;

          if (scope) {
            set(currentState, scope, {
              ...get(currentState, scope),
              ...newState,
            });
          } else {
            currentState = {
              ...currentState,
              ...newState,
            };
          }
        });

        emitter.emit('update', {
          newState: currentState,
          scope: actionScope,
        });
      };
    }),
    emitter,
  ];
};
