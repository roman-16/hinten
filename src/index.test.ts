import { act, renderHook } from '@testing-library/react-hooks';
import createStore from '.';

const useRetext = createStore(global.store);

it('updates the state on dispatch', () => {
  const { result } = renderHook(useRetext);

  act(() => {
    result.current.dispatch.setCount(100);
    result.current.dispatch.increment();
    result.current.dispatch.sideMenu.increment();
  });

  expect(result).toMatchSnapshot();
});

it('selects the nested state', () => {
  const { result: outerHook } = renderHook(useRetext);
  const { result: innerHook } = renderHook(() => useRetext('sideMenu'));

  act(() => {
    outerHook.current.dispatch.increment();
    innerHook.current.dispatch.increment();
    innerHook.current.dispatch.toggle();
  });

  expect(outerHook).toMatchSnapshot();
  expect(innerHook).toMatchSnapshot();
});
