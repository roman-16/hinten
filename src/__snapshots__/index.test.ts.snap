// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selects the nested state 1`] = `
Object {
  "current": Object {
    "dispatch": Object {
      "increment": [Function],
      "setCount": [Function],
      "sideMenu": Object {
        "child": Object {
          "toggle": [Function],
        },
        "increment": [Function],
        "toggle": [Function],
      },
    },
    "state": Object {
      "count": 102,
      "sideMenu": Object {
        "child": Object {
          "isExpanded": true,
        },
        "isOpen": true,
        "maxItems": 104,
      },
    },
  },
  "error": undefined,
}
`;

exports[`selects the nested state 2`] = `
Object {
  "current": Object {
    "dispatch": Object {
      "child": Object {
        "toggle": [Function],
      },
      "increment": [Function],
      "toggle": [Function],
    },
    "state": Object {
      "child": Object {
        "isExpanded": true,
      },
      "isOpen": true,
      "maxItems": 104,
    },
  },
  "error": undefined,
}
`;

exports[`updates the state on dispatch 1`] = `
Object {
  "current": Object {
    "dispatch": Object {
      "increment": [Function],
      "setCount": [Function],
      "sideMenu": Object {
        "child": Object {
          "toggle": [Function],
        },
        "increment": [Function],
        "toggle": [Function],
      },
    },
    "state": Object {
      "count": 101,
      "sideMenu": Object {
        "child": Object {
          "isExpanded": false,
        },
        "isOpen": false,
        "maxItems": 102,
      },
    },
  },
  "error": undefined,
}
`;
